std/experimental/allocator/package.d

```diff
@@ -1123,48 +1123,107 @@ allocated with the same allocator.
 */
 void dispose(A, T)(auto ref A alloc, T* p)
 {
+    if (!p) return;
     static if (hasElaborateDestructor!T)
-    {
-        destroy(*p);
-    }
+        p.__dtor;
     alloc.deallocate((cast(void*)p)[0 .. T.sizeof]);
 }
 
 /// Ditto
 void dispose(A, T)(auto ref A alloc, T p)
-if (is(T == class) || is(T == interface))
+if (is(T == class) && T.stringof != Object.stringof)
+{
+    if (!p) return;
+    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
+    static if (__traits(hasMember, T, "__dtor"))
```

Comment:
Otherwise the overload that handles the `@nogc` destructor would look like this

```D
void dispose(A, T)(auto ref A alloc, T p)
if (is(T == class) && T.stringof != Object.stringof)
{
    if (!p) return;
    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
    static if (__traits(hasMember, T, "__dtor"))
        p.__dtor;
    else
    {
        TypeInfo_Class tic = cast(TypeInfo_Class) typeid(p);
        void* dtorPtr = tic.destructor;
        while (!dtorPtr && tic.base)
        {
            tic = tic.base;
            dtorPtr = tic.destructor;
        }
        if (tic && tic.destructor)
        {
            dispose(alloc, cast(Object) p);
            return;
        }
    }
    alloc.deallocate(support);
}�
```

not good.